dali/operators/math/expressions/expression_impl_factory.h

```diff
@@ -38,9 +38,83 @@ namespace dali {
 
 #define ALLOWED_BIN_OPS                                                                            \
   (ArithmeticOp::add, ArithmeticOp::sub, ArithmeticOp::mul, ArithmeticOp::div, ArithmeticOp::fdiv, \
-      ArithmeticOp::mod, ArithmeticOp::eq, ArithmeticOp::neq, ArithmeticOp::lt, ArithmeticOp::leq, \
-      ArithmeticOp::gt, ArithmeticOp::geq, ArithmeticOp::bit_and, ArithmeticOp::bit_or,            \
-      ArithmeticOp::bit_xor)
+  ArithmeticOp::mod, ArithmeticOp::min, ArithmeticOp::max, ArithmeticOp::eq, ArithmeticOp::neq,    \
+  ArithmeticOp::lt, ArithmeticOp::leq, ArithmeticOp::gt, ArithmeticOp::geq,                        \
+  ArithmeticOp::bit_and, ArithmeticOp::bit_or, ArithmeticOp::bit_xor)
+
+#define ALLOWED_TERNARY_OPS \
+  (ArithmeticOp::clamp)
+
+namespace expression_detail {
+
+/**
+ * @brief Pass through as a pointer to T or return the pointed value based on `as_ptr`
+ */
+template <bool as_ptr, typename T>
+DALI_HOST_DEV std::enable_if_t<as_ptr, const T*> Pass(const T* ptr) {
+  return ptr;
+}
+
+/**
+ * @brief Pass through as a pointer to T or return the pointed value based on `as_ptr`
+ */
+template <bool as_ptr, typename T>
+DALI_HOST_DEV std::enable_if_t<!as_ptr, T> Pass(const T* ptr) {
+  return *ptr;
+}
+
+/**
+ * @brief Pass through as a `const void *` or return the pointed value cast from `type_id` to T
+ *        based on `as_ptr`
+ */
+template <bool as_ptr, typename T>
+DALI_HOST_DEV std::enable_if_t<as_ptr, const void*> Pass(const void* ptr, DALIDataType) {
+  return ptr;
+}
+
+
+/**
+ * @brief Pass through as a `const void *` or return the pointed value cast from `type_id` to T
+ *        based on `as_ptr`
+ */
+template <bool as_ptr, typename T>
+DALI_HOST_DEV std::enable_if_t<!as_ptr, T> Pass(const void* ptr, DALIDataType type_id) {
+  TYPE_SWITCH(type_id, type2id, AccessType, ARITHMETIC_ALLOWED_TYPES, (
+    const auto *access = reinterpret_cast<const AccessType*>(ptr);
+    return static_cast<T>(*access);
+  ), return {}; );  // NOLINT(whitespace/parens)
+}
+
+template <typename T>
+DALI_HOST_DEV T Access(const T* ptr, int64_t idx) {
+  return ptr[idx];
+}
+
+template <typename T>
+DALI_HOST_DEV T Access(T value, int64_t) {
+  return value;
+}
+
+template <typename T>
+DALI_HOST_DEV T Access(const void* ptr, int64_t idx, DALIDataType type_id) {
+  TYPE_SWITCH(type_id, type2id, AccessType, ARITHMETIC_ALLOWED_TYPES, (
+    const auto *access = reinterpret_cast<const AccessType*>(ptr);
+    return static_cast<T>(access[idx]);
+  ), return {}; );  // NOLINT(whitespace/parens)
+}
+
+template <typename T>
+DALI_HOST_DEV T Access(T value, int64_t, DALIDataType) {
+  return value;
+}
+
+template <bool as_ptr, typename T>
+using param_t = std::conditional_t<as_ptr, const T*, T>;
+
+template <bool as_ptr, typename T>
+using param2_t = std::conditional_t<as_ptr, const void*, T>;
```

Comment:
What is this 2 ? Can we have a meaningful name?

Reply:
Again, leftover, I fixed the functions and forgot about the alias, sorry.